### PR TITLE
Prevent creating unnamed policy when loading definition

### LINKED ISSUE #971: Handle or forbid unnamed runtime parameters
Apparently, it's possible to create an unnamed policy according to this mailing-list thread:
https://groups.google.com/d/topic/rabbitmq-users/o_2zx3E2jN8/discussion

The name of the parameter is then set to `undefined` and is impossible to modify/delete.

We should either prevent that or give an implicit name. The former is probably the preferred method.

### PR DESCRIPTION
**Pull request by polaris-alioth**
## Proposed Changes

Fixes #971
Prevent creating unnamed policy when loading definition

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [x] Bug fix (non-breaking change which fixes issue #NNNN)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause an observable behavior change in existing systems)
- [ ] Documentation improvements (corrections, new content, etc)
- [ ] Cosmetic change (whitespace, formatting, etc)
- [ ] Build system and/or CI

## Checklist

_Put an `x` in the boxes that apply.
You can also fill these out after creating the PR.
If you're unsure about any of them, don't hesitate to ask on the mailing list.
We're here to help!
This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the `CONTRIBUTING.md` document
- [x] I have signed the CA (see https://cla.pivotal.io/sign/rabbitmq)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All tests pass locally with my changes
- [x] If relevant, I have added necessary documentation to https://github.com/rabbitmq/rabbitmq-website
- [x] If relevant, I have added this change to the first version(s) in release-notes that I expect to introduce it

## Further Comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc.

### CODE REVIEW

File: deps/rabbit/src/rabbit_definitions.erl [hunk 1]
```diff
@@ -464,6 +464,12 @@ add_policy(Param, Username) ->
 
 add_policy(VHost, Param, Username) ->
     Key   = maps:get(name,  Param, undefined),
+    case Key == undefined of
```
Comment (michaelklishin) [inline deps/rabbit/src/rabbit_definitions.erl:467]: In idiomatic Erlang, you usually match on `undefined`, not compare to it.

File: deps/rabbit/src/rabbit_definitions.erl [hunk 2]
```diff
@@ -464,6 +464,12 @@ add_policy(Param, Username) ->
...
+      true ->
+        rabbit_misc:protocol_error(
```
Comment (michaelklishin) [inline deps/rabbit/src/rabbit_definitions.erl:469]: While this ends up calling `erlang:exit/1` and technically might be how other errors are handled in this function,
`rabbit_misc:protocol_error/3` is mean to be used to terminate a channel and return an AMQP 0-9-1 error back to the client. It's therefore not appropriate here.

Comment (michaelklishin) [reply deps/rabbit/src/rabbit_definitions.erl:469]: Please use `exit/1` directly as the same function does if `rabbit_policy:set/7` returns an error.

**Review by michaelklishin — CHANGES_REQUESTED** [review]
Thank you.

You are on the right track but this needs some more work.

File: deps/rabbit/src/rabbit_definitions.erl (hunk 1 above)
Comment (polaris-alioth) [reply deps/rabbit/src/rabbit_definitions.erl:467]: Thank you for your advise.
Belowing is my modified code
```
    Key   = maps:get(name,  Param, undefined),
    case Key of
      undefined -> exit("policy name not defined");
      _ -> ok
    end,
```

Comment (michaelklishin) [reply deps/rabbit/src/rabbit_definitions.erl:467]: You can push to your branch and the PR will be updated.

Comment (polaris-alioth) [reply deps/rabbit/src/rabbit_definitions.erl]: I have push this to my branch. The  PR is updated.

**Comment by michaelklishin** [comment]
Backported manually to `v3.9.x` (for `3.9.12`) and `v3.8.x` (for `3.8.27`). Thank you.

**Comment by polaris-alioth** [comment]
Sorry, When I saw this message, this PR has been backported to v3.9.x and v3.8.x.

> Backported manually to `v3.9.x` (for `3.9.12`) and `v3.8.x` (for `3.8.27`). Thank you.

**Pull request merged**